### PR TITLE
Improve neuron model

### PR DESCRIPTION
**Pull request by mundya**
Removes the LFSR approximation for inter-step spiking and instead uses
the overshoot to determine the voltage on leaving the refractory period.

Before (using the LFSR) (sadly this plot doesn't show the variance, which is HUGE):
before

After:
after

(@tcstewar, I can't see the colours apart either, but you get the idea...)

### CODE REVIEW

**Comment by mundya** [comment]
Bonus graphs, this model drives the ensemble into saturation:

``` python
with nengo.Network(seed=2) as model:
    in_node = nengo.Node([1.0, -1.0])
    ens = nengo.Ensemble(10, 2)

    nengo.Connection(in_node, ens)

    probe = nengo.Probe(ens, synapse=0.05)
```

(The outer pair of lines are SpiNNaker, the inner are Nengo).  Before this PR (note the noisiness):
saturated_before

After this PR (same ordering):
saturated_after

Not identical to Nengo, but definitely improved.

File: spinnaker_components/ensemble/ensemble_update.c [hunk 1]
```diff
@@ -57,43 +55,36 @@ void ensemble_update(uint ticks, uint arg1) {
...
-
-    v_voltage += v_delta;
+    v_delta = (i_membrane - v_voltage) * g_ensemble.exp_dt_over_t_rc;
 
     // Voltages can't go below 0.0
-    if( v_voltage < 0.0k ) {
+    v_voltage += v_delta;
+    if(v_voltage < 0.0k)
+    {
       v_voltage = 0.0k;
```
Comment (tcstewar) [inline spinnaker_components/ensemble/ensemble_update.c:64]: would this be slightly better as 

``` c
if(v_voltage < 0.0k)
{
    v_voltage = v_delta < 0.0k ? 0.0k : v_delta;
}
```

File: spinnaker_components/ensemble/ensemble.h [hunk 2]
```diff
@@ -65,10 +58,11 @@ typedef struct ensemble_parameters {
   uint machine_timestep;   //!< Machine time step  / useconds
 
   uint t_ref;              //!< Refractory period \f$\tau_{ref} - 1\f$ / steps
-  value_t dt_over_t_rc;    //!< \f$\frac{dt}{\tau_{rc}}\$
+  value_t exp_dt_over_t_rc;    //!< \f$-\exp{\frac{dt}{\tau_{rc}}}\$
 
   current_t *i_bias;        //!< Population biases \f$1 \times N\f$
-  neuron_status_t *status;  //!< Neuron status
+  value_t *neuron_voltage;  //!< Neuron voltages
+  uint8_t *neuron_refractory;  //!< Refractory states
```
Comment (tcstewar) [inline spinnaker_components/ensemble/ensemble.h:65]: Does this change affect the formula for determining how many neuron there can be on a core?

File: nengo_spinnaker/operators/lif.py [hunk 3]
```diff
@@ -259,8 +264,8 @@ def write_subregion_to_file(self, fp, vertex_slice):
             self.n_output_dimensions,
             n_neurons,
             self.machine_timestep,
-            int(self.t_ref // self.dt),
-            tp.value_to_fix(self.dt / self.t_rc),
+            int(self.t_ref // self.dt),  # tau_ref expressed as in ticks
+            tp.value_to_fix(-np.expm1(-self.dt / self.t_rc) * (1.0 - 2**-17)),
```
Comment (tcstewar) [inline nengo_spinnaker/operators/lif.py]: I'm relatively sure that math is right, but it'd be nice to have a plot of what happens to the tuning curve as you change that fudge factor.

File: spinnaker_components/ensemble/ensemble_update.c (hunk 1 above)
Comment (mundya) [inline spinnaker_components/ensemble/ensemble_update.c:64]: I don't think so, the case where 0 < |v_delta| < |v_voltage| would work differently...

File: spinnaker_components/ensemble/ensemble.h (hunk 2 above)
Comment (mundya) [inline spinnaker_components/ensemble/ensemble.h:65]: By up to 3 bytes, so I'm not overly worried about that.  There's already some looseness in the calculation.

File: spinnaker_components/ensemble/ensemble_update.c (hunk 1 above)
Comment (tcstewar) [inline spinnaker_components/ensemble/ensemble_update.c:64]: Ah, I see...  yup, that now makes sense to me.  (and it's exactly the same as it used to be)

File: nengo_spinnaker/operators/lif.py (hunk 3 above)
Comment (mundya) [inline nengo_spinnaker/operators/lif.py]: So after playing around with different fudge factors 1 - 2**-11 seems gets pretty close to the ideal tuning curve... I have data if anyone wants to spend time glancing at very small tuning curve differences.

Comment (mossblaser) [inline nengo_spinnaker/operators/lif.py]: As an outsider I don't have an intuition for why there is a fudge factor
and what it is correcting. Is out simple enough to document inline in the
code here?

Comment (mundya) [inline nengo_spinnaker/operators/lif.py]: Please look to the top of the function.

File: spinnaker_components/ensemble/ensemble.h (hunk 2 above)
Comment (tcstewar) [inline spinnaker_components/ensemble/ensemble.h:65]: According to @mundya it's 5 bytes per neuron.  :)

**Comment by mundya** [comment]
I've added in the 5-bytes per neuron of DTCM usage and hopefully better explained the fudge factor.  Comments welcomed :)

**Comment by tcstewar** [comment]
Looks good to me!

File: spinnaker_components/ensemble/ensemble.h (hunk 2 above)
Comment (neworderofjamie) [inline spinnaker_components/ensemble/ensemble.h:65]: I spotted this structure during yesterdays C-meddling and purely from a performance P.O.V., I think bit field structures should be avoided. In terms of neurons we're CPU, not memory bound and a lot of sign-extending, shifting and masking horror'll emerge from updating those fields.

Comment (mundya) [inline spinnaker_components/ensemble/ensemble.h:65]: Sorry, @neworderofjamie, which structure? I'm assuming that the two arrays is fine?

Comment (neworderofjamie) [inline spinnaker_components/ensemble/ensemble.h:65]: The bit-field that was there before with the 28-bits and the 4-bits - My
comment was more a late-to-the-party affirmation of your change than
anything else!

Comment (mundya) [inline spinnaker_components/ensemble/ensemble.h:65]: > The bit-field that was there before with the 28-bits and the 4-bits - My comment was more a late-to-the-party affirmation of your change than anything else!

Thanks!  We should reprofile at some point, I suspect the cost per dimension has dropped fairly significantly.

**Pull request merged**